File: config/src/test/scala/net/kemitix/thorp/config/ConfigurationBuilderTest.scala
package net.kemitix.thorp.config

import java.nio.file.{Path, Paths}
import java.util

import net.kemitix.thorp.domain.Filter.{Exclude, Include}
import net.kemitix.thorp.domain._
import net.kemitix.thorp.filesystem.TemporaryFolder
import org.scalatest.FunSpec

import scala.jdk.CollectionConverters._

class ConfigurationBuilderTest extends FunSpec with TemporaryFolder {

  private val pwd: Path              = Paths.get(System.getenv("PWD"))
  private val aBucket                = Bucket.named("aBucket")
  private val coBucket: ConfigOption = ConfigOption.bucket(aBucket.name)
  private val thorpConfigFileName    = ".thorp.conf"

  private def configOptions(options: ConfigOption*): ConfigOptions =
    ConfigOptions.create(
      (List[ConfigOption](
        ConfigOption.ignoreUserOptions(),
        ConfigOption.ignoreGlobalOptions()
      ) ++ options).asJava)

  describe("when no source") {
    it("should use the current (PWD) directory") {
      val expected = Right(Sources.create(List(pwd).asJava))
      val options  = configOptions(coBucket)
      val result   = invoke(options).sources
      assertResult(expected)(result)
    }
  }
  describe("a source") {
    describe("with .thorp.conf") {
      describe("with settings") {
        withDirectory(source => {
          createFile(source,
                     thorpConfigFileName,
                     util.Arrays.asList("bucket = a-bucket",
                                        "prefix = a-prefix",
                                        "include = an-inclusion",
                                        "exclude = an-exclusion"))
          val result = invoke(configOptions(ConfigOption.source(source)))
          it("should have bucket") {
            val expected = Right(Bucket.named("a-bucket"))
            assertResult(expected)(result.bucket)
          }
          it("should have prefix") {
            val expected = Right(RemoteKey.create("a-prefix"))
            assertResult(expected)(result.prefix)
          }
          it("should have filters") {
            val expected =
              Right(
                List[Filter](Exclude.create("an-exclusion"),
                             Include.create("an-inclusion")))
            assertResult(expected)(result.filters)
          }
        })
      }
    }
  }
  describe("when has a single source with no .thorp.conf") {
    it("should only include the source once") {
      withDirectory(aSource => {
        val expected = Right(Sources.create(List(aSource).asJava))
        val options  = configOptions(ConfigOption.source(aSource), coBucket)
        val result   = invoke(options).sources
        assertResult(expected)(result)
      })
    }
  }
  describe("when has two sources") {
    it("should include both sources in order") {
      withDirectory(currentSource => {
        withDirectory(previousSource => {
          val expected = Right(List(currentSource, previousSource))
          val options = configOptions(ConfigOption.source(currentSource),
                                      ConfigOption.source(previousSource),
                                      coBucket)
          val result = invoke(options).sources.paths
          assertResult(expected)(result)
        })
      })
    }
  }
  describe("when current source has .thorp.conf with source to another") {
    it("should include both sources in order") {
      withDirectory(currentSource => {
        withDirectory(previousSource => {
          createFile(currentSource,
                     thorpConfigFileName,
                     util.Arrays.asList(s"source = $previousSource"))
          val expected = Right(List(currentSource, previousSource))
          val options =
            configOptions(ConfigOption.source(currentSource), coBucket)
          val result = invoke(options).sources.paths
          assertResult(expected)(result)
        })
      })
    }
    describe("when settings are in current and previous") {
      it("should include settings from only current") {
        withDirectory(previousSource => {
          withDirectory(currentSource => {
            createFile(
              currentSource,
              thorpConfigFileName,
              util.Arrays.asList(s"source = $previousSource",
                                 "bucket = current-bucket",
                                 "prefix = current-prefix",
                                 "include = current-include",
                                 "exclude = current-exclude")
            )
            createFile(
              previousSource,
              thorpConfigFileName,
              util.Arrays.asList("bucket = previous-bucket",
                                 "prefix = previous-prefix",
                                 "include = previous-include",
                                 "exclude = previous-exclude")
            )
            // should have both sources in order
            val expectedSources =
              Right(Sources.create(List(currentSource, previousSource).asJava))
            // should have bucket from current only
            val expectedBuckets = Right(Bucket.named("current-bucket"))
            // should have prefix from current only
            val expectedPrefixes = Right(RemoteKey.create("current-prefix"))
            // should have filters from both sources
            val expectedFilters = Right(
              List[Filter](Filter.Exclude.create("current-exclude"),
                           Filter.Include.create("current-include")))
            val options = configOptions(ConfigOption.source(currentSource))
            val result  = invoke(options)
            assertResult(expectedSources)(result.sources)
            assertResult(expectedBuckets)(result.bucket)
            assertResult(expectedPrefixes)(result.prefix)
            assertResult(expectedFilters)(result.filters)
          })
        })
      }
    }
  }

  describe(
    "when source has thorp.config source to another source that does the same") {
    it("should only include first two sources") {
      withDirectory(currentSource => {
        withDirectory(parentSource => {
          createFile(currentSource,
                     thorpConfigFileName,
                     util.Arrays.asList(s"source = $parentSource"))
          withDirectory(grandParentSource => {
            createFile(parentSource,
                       thorpConfigFileName,
                       util.Arrays.asList(s"source = $grandParentSource"))
            val expected = Right(List(currentSource, parentSource))
            val options =
              configOptions(ConfigOption.source(currentSource), coBucket)
            val result = invoke(options).sources.paths
            assertResult(expected)(result)
          })
        })
      })
    }
  }

  private def invoke(configOptions: ConfigOptions) =
    ConfigurationBuilder.buildConfig(configOptions)

}

File: config/src/test/java/net/kemitix/thorp/config/ConfigurationBuilderTest.java
package net.kemitix.thorp.config;

import net.kemitix.thorp.domain.Bucket;
import net.kemitix.thorp.domain.RemoteKey;
import net.kemitix.thorp.filesystem.TemporaryFolder;
import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ConfigurationBuilderTest
        implements WithAssertions {
    Path pwd = Paths.get(System.getenv("PWD"));
    Bucket aBucket = Bucket.named("aBucket");
    ConfigOption coBucket = ConfigOption.bucket(aBucket.name());
    String thorpConfigFileName = ".thorp.conf";
    ConfigOptions configOptions(List<ConfigOption> options) {
        List<ConfigOption> optionList = new ArrayList<>(options);
        optionList.add(ConfigOption.ignoreUserOptions());
        optionList.add(ConfigOption.ignoreGlobalOptions());
        return ConfigOptions.create(optionList);
    }
    @Test
    @DisplayName("when no source then user current directory")
    public void whenNoSource_thenUseCurrentDir() throws IOException, ConfigValidationException {
        Configuration result = ConfigurationBuilder.buildConfig(
                configOptions(Collections.singletonList(coBucket)));
        assertThat(result.sources.paths()).containsExactly(pwd);
    }
    @Nested
    @DisplayName("default source")
    public class DefaultSourceTests {
        @Nested
        @DisplayName("with .thorp.conf")
        public class WithThorpConfTests implements TemporaryFolder {
            @Test
            @DisplayName("with settings")
            public void WithSettingsTests() {
                withDirectory(source -> {
                    //given
                    List<String> settings = Arrays.asList(
                            "bucket = a-bucket",
                            "prefix = a-prefix",
                            "include = an-inclusion",
                            "exclude = an-exclusion"
                    );
                    createFile(source, thorpConfigFileName, settings);
                    //when
                    Configuration result =
                            invoke(configOptions(Collections.singletonList(
                                    ConfigOption.source(source))));
                    //then
                    assertThat(result.bucket).isEqualTo(Bucket.named("a-bucket"));
                    assertThat(result.prefix).isEqualTo(RemoteKey.create("a-prefix"));
                    assertThat(result.filters).hasSize(2)
                            .anySatisfy(filter ->
                                    assertThat(filter.predicate()
                                            .test("an-exclusion")).isTrue())
                            .anySatisfy(filter ->
                                    assertThat(filter.predicate()
                                            .test("an-inclusion")).isTrue());
                });
            }
        }
    }
    @Nested
    @DisplayName("single source")
    public class SingleSourceTests implements TemporaryFolder {
        @Test
        @DisplayName("has single source")
        public void hasSingleSource() {
            withDirectory(aSource -> {
                Configuration result =
                        invoke(
                                configOptions(Arrays.asList(
                                        ConfigOption.source(aSource),
                                        coBucket)));
                assertThat(result.sources.paths()).containsExactly(aSource);
            });
        }
    }
    @Nested
    @DisplayName("multiple sources")
    public class MultipleSources implements TemporaryFolder {
        @Test
        @DisplayName("included in order")
        public void hasBothSourcesInOrder() {
            withDirectory(currentSource -> {
                withDirectory(previousSource -> {
                    Configuration result =
                            invoke(configOptions(Arrays.asList(
                                    ConfigOption.source(currentSource),
                                    ConfigOption.source(previousSource),
                                    coBucket)));
                    assertThat(result.sources.paths())
                            .containsExactly(
                                    currentSource,
                                    previousSource);
                });
            });
        }
    }

    @Nested
    @DisplayName("config file includes another source")
    public class ConfigLinkedSourceTests implements TemporaryFolder {
        @Test
        @DisplayName("include the linked source")
        public void configIncludeOtherSource() {
            withDirectory(currentSource -> {
                withDirectory(previousSource -> {
                    createFile(currentSource,
                            thorpConfigFileName,
                            Collections.singletonList(
                                    "source = " + previousSource));
                    Configuration result = invoke(configOptions(Arrays.asList(
                            ConfigOption.source(currentSource),
                            coBucket)));
                    assertThat(result.sources.paths())
                            .containsExactly(
                                    currentSource,
                                    previousSource);
                });
            });
        }

        @Test
        @DisplayName("when linked source has config file")
        public void whenSettingsFileInBothSources() {
            withDirectory(currentSource -> {
                withDirectory(previousSource -> {
                    //given
                    createFile(currentSource,
                            thorpConfigFileName,
                            Arrays.asList(
                                    "source = " + previousSource,
                                    "bucket = current-bucket",
                                    "prefix = current-prefix",
                                    "include = current-include",
                                    "exclude = current-exclude"));
                    createFile(previousSource,
                            thorpConfigFileName,
                            Arrays.asList(
                                    "bucket = previous-bucket",
                                    "prefix = previous-prefix",
                                    "include = previous-include",
                                    "exclude = previous-exclude"));
                    //when
                    Configuration result = invoke(configOptions(Arrays.asList(
                            ConfigOption.source(currentSource),
                            coBucket)));
                    //then
                    assertThat(result.sources.paths()).containsExactly(currentSource, previousSource);
                    assertThat(result.bucket.name()).isEqualTo("current-bucket");
                    assertThat(result.prefix.key()).isEqualTo("current-prefix");
                    assertThat(result.filters).anyMatch(filter -> filter.predicate().test("current-include"));
                    assertThat(result.filters).anyMatch(filter -> filter.predicate().test("current-exclude"));
                    assertThat(result.filters).noneMatch(filter -> filter.predicate().test("previous-include"));
                    assertThat(result.filters).noneMatch(filter -> filter.predicate().test("previous-exclude"));
                });
            });
        }
    }
    @Nested
    @DisplayName("linked source links to third source")
    public class LinkedSourceLinkedSourceTests implements TemporaryFolder {
        @Test
        @DisplayName("ignore third source")
        public void ignoreThirdSource() {
            withDirectory(currentSource -> {
                withDirectory(parentSource -> {
                    createFile(currentSource, thorpConfigFileName,
                            Collections.singletonList("source = " + parentSource));
                    withDirectory(grandParentSource -> {
                        createFile(parentSource, thorpConfigFileName,
                                Collections.singletonList("source = " + grandParentSource));
                        //when
                        Configuration result = invoke(configOptions(Arrays.asList(
                                ConfigOption.source(currentSource), coBucket)));
                        //then
                        assertThat(result.sources.paths())
                                .containsExactly(currentSource, parentSource)
                                .doesNotContain(grandParentSource);
                    });
                });
            });
        }
    }

    @Test
    @DisplayName("when batch mode option then batch mode in configuration")
    public void whenBatchMode_thenBatchMode() {
        Configuration result= invoke(configOptions(Arrays.asList(
                ConfigOption.batchMode(),
                coBucket)));
        assertThat(result.batchMode).isTrue();
    }

    public Configuration invoke(ConfigOptions configOptions) {
        try {
            return ConfigurationBuilder.buildConfig(configOptions);
        } catch (IOException | ConfigValidationException e) {
            throw new RuntimeException(e);
        }
    }
}
